craft_application/application.py

```diff
@@ -308,9 +308,10 @@ def get_project(
 
         build_on = host_arch
 
+        yaml_data = self._transform_project_yaml(yaml_data, build_on, build_for)
```

Comment:
This won't work because we need the environment that craft-parts creates in order to do the transform.

What's the use case for transforming the YAML before setting up partitions anyway? The only thing i can think of would be if we want to create the set of partitions based on a transform of the project data, but the way to do this is to extract that particular transform out into a function and run it twice - once in `extra_yaml_transform` and once in `_setup_partitions`.

Reply:
Ok that's fair.  So then which of the other options would you be leaning towards?

Reply:
In the application, have a function that does the minimal transform needed to support both `_setup_partitions` and `_extra_yaml_transform`, ensuring that it returns a new dict rather than modifying the original one (do a `copy.deepcopy(yaml_data)` if needed).

Then you can call that from your app's `_setup_partitions` as well as its `_extra_yaml_transform`.

Reply:
The code in question in the application isn't doing a transform.  It's creating an object on `self` that's dependent on `yaml_data` and `build_for`.  This object is needed by `_setup_partitions` (as well as `_extra_yaml_transform`, and `_configure_services`).

In `craft_application.Application.get_project`, there's nothing I can hook into after `build_for` is determined, but before the call to `_setup_partitions`.  So either I need to get `build_for` into `_setup_partitions` (option # 2 above), or add a new function I can implement as a hook before `_setup_partitions` gets called (option # 3 above).  (Or the original "function reordering" implementation which you said wouldn't work.)  Or something else I haven't thought of.